janus_server/src/aggregator.rs

```diff
@@ -1091,11 +1086,68 @@ impl VdafOps {
+        //  * could cause us to fail to meet the minimum batch size for the task,
+        //  * or for particularly pathological timing, could cause us to aggregate a different set
+        //    of reports than the leader did (though the checksum will detect this).
+        //
+        // There's not much the helper can do about this, because an aggregate job might be
+        // unfinished because it's waiting on an aggregate sub-protocol message that is never coming
+        // because the leader has abandoned that job.
+        //
+        // Thus the helper has no choice but to assume that any unfinished aggregation jobs were
+        // intentionally abandoned by the leader and service the aggregate share request with
+        // whatever batch unit aggregations are available now.
+        //
+        // See issue #104 for more discussion.
+
         let mut total_report_count = 0;
         let mut total_checksum = [0u8; 32];
         let mut total_aggregate_share: Option<A::AggregateShare> = None;
 
         for batch_unit_aggregation in &batch_unit_aggregations {
+            // Check how many rows in aggregate_share_jobs have a batch interval that includes this
+            // batch unit. Each such row consumes one unit of batch lifetime (§4.6).
+            //
+            // We have to check each batch unit interval separately rather than checking once how
+            // many times aggregate_share_req.batch_interval overlaps with any row. Suppose we had:
+            //
+            //   * task.max_batch_lifetime = 2,
+            //   * an AggregateShareReq.batch interval that spans two batch units,
+            //   * and that each of those batch units has been collected once before.
+            //
+            // A further AggregateShareReq including either or both of the units is permissible, but
+            // if we queried how many rows overlap with that interval, we would get 2 and refuse the
+            // request. We must check the unit intervals individually to notice that each has enough
+            // remaining lifetime to permit the share request.
+            let unit_interval = Interval::new(
+                batch_unit_aggregation.unit_interval_start,
+                task.min_batch_duration,
+            )?;
+            let consumed_batch_lifetime = tx
+                .get_aggregate_share_job_count_by_batch_unit(task.id, unit_interval)
+                .await?;
+            if consumed_batch_lifetime == task.max_batch_lifetime {
+                warn!(
+                    ?task.id, ?aggregate_share_req, ?unit_interval,
+                    "refusing aggregate share request because lifetime for batch unit has been consumed"
+                );
+                return Err(datastore::Error::User(
+                    Error::BatchLifetimeExceeded(task.id).into(),
+                ));
+            }
+            if consumed_batch_lifetime > task.max_batch_lifetime {
```

Comment:
Hypothetically, this branch could be taken if we modified a Task to lower its `max_batch_lifetime`. It may still be useful to have a special branch since this is extra-unexpected, but for that reason I'm somewhat wary of panicking.

Reply:
I think that some task parameters are effectively immutable (e.g. anything distributed to clients), but other parameters might be mutable -- but I definitely agree that's something to be considered for DAP.

With that in mind, I filed https://github.com/ietf-wg-ppm/ppm-specification/issues/237. Moving forward with immutable tasks SGTM for now.